### DISCUSSION TITLE
How can I use a Python class decorator in Rust code?

### DISCUSSION THREAD

**Discussion by silvanocerza** [PyO3/pyo3#3537] [Questions]
I have a library written in Python that declares a decorator for classes. Something similar to this:

```python
from types import new_class

class _MyDecorator:
    def __call__(self, cls):
        # Verify it has certain methods...
        return new_class(cls.__name__, cls.__bases__, {"metaclass": MyMetaclass})

decorator = _MyDecorator()
```

And it can be used like this:

```python
@decorator
class Whatever:
    pass
```

So it overwrites the original definition of the `Whatever` class with a new one, containing some new fields, methods, etc.

I know I can call Python functions in Rust, but is it possible to use this kind of decorator in Rust given that it overwrites the class definition?

My intention would be declaring a Python class in Rust like so:

```Rust
#[pyclass]
struct MyClass {
    my_field: i32,
}
```

And decorate it in Rust using `decorator` defined in Python. I tried this and it obviously doesn't work for several reasons.

```Rust
#[pymodule]
#[pyo3(name = "my_extension")]
fn module_with_functions(py: Python<'_>, m: &PyModule) -> PyResult<()> {
    let decorator = py.import("my_decorator.decorator")?;
    let kwargs = PyDict::new(py);
    let res = decorator.call(MyClass, Some(kwargs))?;
    MyClass = res.extract::<MyClass>()?;
    m.add_class::<MyClass>()?;
    Ok(())
}
```

Is there a way to make this work or should I just rewrite the decorator in Rust? 

Another alternative I tried is to wrap the class in Python after it's imported as this seems to work, but it's a bit more annoying.

```python
import my_extension.MyClass
import my_decorator.decorator

MyClass = decorator(MyClass)
```

**Comment by davidhewitt** [comment]
I think we would need to implement this as a new `#[pyclass(decorate_with = "some_function")]`. PR welcome

**Reply by silvanocerza** [reply]
Interesting, I'll give it a try. 🤔

I'll create an issue in any case and refer to this discussion so it doesn't get lost.
